data/Builder.go

```diff
@@ -0,0 +1,72 @@
+package data
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/tidepool-org/platform/validate"
+)
+
+type Datum map[string]interface{}
+
+type DatumArray []Datum
+
+type Builder interface {
+	BuildFromDatum(datum Datum) interface{}
+	BuildFromDatumArray(datumArray DatumArray) ([]interface{}, *validate.ErrorsArray)
+}
+
+type TypeBuilder struct {
+	inject map[string]interface{}
+	Errors *validate.ErrorsArray
+	Index  int
+}
+
+func NewTypeBuilder(inject map[string]interface{}) Builder {
+	return &TypeBuilder{
+		inject: inject,
+		Errors: validate.NewErrorsArray(),
+		Index:  0,
```

Comment:
I think we should not use 'Index', but instead something like 'SourceRoot' as string to be more generic. I can imagine a scenario where we have an embedded array within a datum that we are validating.  Then the 'SourceRoot' would be something like '0/myArray/1', but 'Index' wouldn't work.